src/java.base/share/classes/java/math/BigDecimal.java

```diff
@@ -5219,55 +5215,118 @@ private static boolean needIncrement(MutableBigInteger mdivisor, int roundingMod
         return commonNeedIncrement(roundingMode, qsign, cmpFracHalf, mq.isOdd());
     }
 
+    /**
+     * {@code FIVE_TO_2_TO[n] == 5^(2^n)}
+     */
+    // floor(log2(log5(2^Integer.MAX_VALUE))) == 29
+    private static final BigInteger[] FIVE_TO_2_TO = new BigInteger[30];
+
+    private static int FIVE_TO_2_TO_LEN = 1;
+
+    static {
+        FIVE_TO_2_TO[0] = BigInteger.valueOf(5L);
+    }
+
+    /**
+     * @param n an integer such that {@code 0 <= n < POWERS_OF_5.length}
+     * @return {@code 5^(2^n)}
+     */
+    private static BigInteger fiveToTwoToThe(int n) {
+        if (n >= FIVE_TO_2_TO_LEN) {
+            BigInteger pow = FIVE_TO_2_TO[FIVE_TO_2_TO_LEN - 1];
+            for (int i = FIVE_TO_2_TO_LEN; i <= n; i++)
+                FIVE_TO_2_TO[i] = pow = pow.multiply(pow);
```

Comment:
@archiecobbs AFAIK, the documentation of `BigDecimal` does not specify that the class is thread-safe, but I could be wrong...

Reply:
@fabioromano1, this is a static method manipulating static fields that are shared by all instances; whether individual instances are thread safe is a separate and unrelated issue.

Reply:
The real problem here is the maximal size this table can reach and the time it takes to be filled in the worst case: around 300 MB and several minutes.

Reply:
@rgiulietti Which means, however, wanting to work with a precision of billions of decimal digits, and therefore taking on the consequences...

Reply:
It takes around 100 ms and less than 1 MB to build a table up to 20 rather than 29.
It can be fully constructed in `static { ... }`, avoiding races.

IMO, this would cover the vast majority of the cases encountered in practice.
Other, way less common usages with multi-millions or billions of digits would then impact running time but not _persistent_ memory footprint.

WDYT?

Reply:
@rgiulietti It can be a good compromise between memory and running time...

Reply:
@rgiulietti Though, it's also true that if I want to do astronomical calculations, I probably have a machine that allows me to do so, and so I'd like to be able to make the most of it...

Reply:
`BigDecimal` was not designed to do astronomical computations. One would probably use other libraries for that, e.g., [GMP](https://gmplib.org/) and would not be interested too much in "stripping trailing zeros".
The focus of this class is mainly commercial applications, where decimal arithmetic is sometimes mandated by regulators.

So I'd expect that most usages, for example SQL or commercial application workloads, will process numbers with a few dozens of digits, maybe sometimes some hundreds digits for intermediate results. So I think that accelerating for these sizes makes the most practical sense.

With modest memory and CPU costs, I think even one million digits or so might be processed reasonably fast.
My point is that there's room for improvement in these range of sizes, but perhaps not for larger ones.

More importantly, though, I'd like to avoid one single computation to affect the resident memory footprint of a JVM instance for its entire lifetime.

Reply:
@rgiulietti FYI, I have found that `BigInteger` has already a cache for the repeated squares that uses in radix string conversion, and it is provided by the method `BigInteger.getRadixConversionCache(int radix, int exponent)`. Although, I don't know if it would be a good idea to use it, because it computes all the squares up to `radix^(2^exponent)`, and this may not be necessary but could cost a lot of running time and memory if the number to strip is very large...

Reply:
Seems like `BigInteger`s division either takes care of removing common powers of 2 in the dividend and divisor (Knuth division), or removing them has no practical benefit (Burnikel-Ziegler).

Thus, I'm wondering if using powers of 10 rather than powers of 5 in the stripping logic could simplify the code while retaining comparable performance.

The powers of 10 are already indirectly kept in the class by method `bigTenToThe()` (although they affect resident memory footprint). However, since that method is there to stay and used in other places as well, it might be worthwhile to profit from it.

Reply:
@rgiulietti 
- About powers of two, there is a difference here: all powers of two are removed, not only the common, to use numbers as small as possible.
- About `bigTenToThe()`: its shortcoming is that the powers are calculated multiplying by 10 each time, rather than using repeated squares trick, until the cache completely filled.

These are the main reasons why I decided not to use `bigTenToThe()`.

Reply:
Oh right, `expandBigIntegerTenPowers()` indeed computes and stores a lot of values.

I'm thinking of opening an issue to rewrite it to just keep the values that are actually computed as intermediate results of a repeated squaring implementation.

Reply:
@rgiulietti BTW of opening new issues, I've found that there are some optimizations and code simplifications that can be done in `BigInteger`'s bit operations that uses lowest non-zero word search, like `shiftRight()`, `getLowestSetBit()`, `bitLength()` and `bitCount()`.

Reply:
If you are interested, you can create new PRs without reference to JBS issues, and I'll file them.
As usual, it's easier for review purposes when each PR touches only small groups of strongly related changes rather than big modifications en-bloc.

But you should show JMH benchmark evidence in later comments (not the introductory comment) that there are no performance regressions, in addition to code simplification.

Reply:
What's wrong with the semantics? In my understanding, it should count the ones for non-negative values and the zeros for negative values, where non-negative values have an infinite prefix of zero bits and negative values have an infinite prefix of one bits in two's complement representation.

Note that this is different from `[Long|Integer].bitCount()`, which consider their arguments as finite width unsigned bit strings.

Reply:
@rgiulietti But the documentation is not clear about this, it should be specified...

Reply:
@rgiulietti At least for me, the expression "the bits that differ from its sign bit" is misleading, it seems to implicitly imply that he is talking about the 1s bits which are not sign bits, while it means the bits whose value differs from the value of the sign bits...

Reply:
The class Javadoc mentions, in at least two places, that the two's complement representation conceptually has an infinite number of leading virtual sign bits.
Once this point is clear, IMO the method's own Javadoc is then clear in turn. Maybe it could be improved as in "the bits that differ _in value_ from its sign bit". (Changing the spec requires another process and another, separate issue, but it can be done if needed.)

Reply:
@rgiulietti So it must have been my limitation, but I hadn't realized it, even though I knew that there are virtually infinite sign bits...

Reply:
Turns out that the table maintained by `expandBigIntegerTenPowers()` never stores more than 19 * 16 = 304 entries, totaling some 20 kB of resident storage in the worst case. Nor is the iterative multiplication by 10 a problem for such a small cache.

I tried to reduce the size of the `FIVE_TO_2_TO` array to a length of 16 + 1 instead of 20. The size is then around 80 kB, about the same order of magnitude as for the cache above.
The impact on performance is barely noticeable on large number of trailing zeros, with a slowdown of about 5%.

```
Benchmark                            Mode  Cnt          Score         Error  Units
BigDecimalStripTrailingZeros.testL   avgt   15    2733190.412 ?   51858.195  ns/op
BigDecimalStripTrailingZeros.testM   avgt   15       9262.737 ?      18.105  ns/op
BigDecimalStripTrailingZeros.testS   avgt   15        153.792 ?       1.234  ns/op
BigDecimalStripTrailingZeros.testXL  avgt   15  466326466.667 ? 6076783.818  ns/op
BigDecimalStripTrailingZeros.testXS  avgt   15         21.336 ?       0.058  ns/op
Finished running test 'micro:java.math.BigDecimalStripTrailingZeros'
```